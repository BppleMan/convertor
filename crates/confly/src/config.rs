use convertor::config::Config;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

mod client_config;

pub use client_config::*;
use convertor::config::proxy_client::ProxyClient;

#[derive(Debug, Clone)]
#[derive(Serialize, Deserialize)]
pub struct ConflyConfig {
    #[serde(flatten)]
    pub common: Config,

    #[serde(flatten)]
    pub clients: HashMap<ProxyClient, ClientConfig>,
}
