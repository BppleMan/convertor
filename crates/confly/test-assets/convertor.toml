secret = "bppleman"
server = "http://127.0.0.1:8080/"

[subscription]
sub_url = "http://127.0.0.1:8080/subscription"
interval = 86400
strict = true

[subscription.headers]
User-Agent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML), like Gecko) Chrome/"
sec-ch-ua = '"Not)A;Brand";v="8", "Chromium";v="138", "Google Chrome";v="138"'

[surge]
config_dir = "surge"
main_profile = "surge.conf"
raw = "raw.conf"
raw_profile = "raw_profile.conf"
rules = "rules.dconf"

[clash]
config_dir = "clash"
main_profile = "config.yaml"
