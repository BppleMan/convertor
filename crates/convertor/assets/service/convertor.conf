server {
    listen 443 ssl;
    http2 on;
    server_name convertor.bppleman.cn;

    # SSL 证书
    ssl_certificate     /etc/ssl/certs/bppleman.cn.pem;
    ssl_certificate_key /etc/ssl/private/bppleman.cn.key;

    # SSL 加密推荐配置
    ssl_protocols TLSv1.2 TLSv1.3;
    ssl_prefer_server_ciphers on;
    ssl_ciphers HIGH:!aNULL:!MD5;

    # 专属日志路径
    access_log /var/log/nginx/convertor_access.log main;
    error_log  /var/log/nginx/convertor_error.log warn;

    # 反代设置
    location / {
        proxy_pass http://127.0.0.1:8080;

        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        proxy_buffering off;

        proxy_connect_timeout 10s;
        proxy_send_timeout    30s;
        proxy_read_timeout    30s;
    }

    # gzip 压缩
    gzip on;
    gzip_types text/plain application/json application/javascript text/css;

    # 安全 header（可按需删改）
    add_header X-Content-Type-Options nosniff;
    add_header X-Frame-Options DENY;
    add_header X-XSS-Protection "1; mode=block";
}
