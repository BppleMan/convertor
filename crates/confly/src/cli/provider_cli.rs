use clap::Args;
use color_eyre::Result;
use color_eyre::eyre::eyre;
use color_eyre::owo_colors::OwoColorize;
use convertor::common::config::ConvertorConfig;
use convertor::common::config::provider_config::Provider;
use convertor::common::config::proxy_client_config::ProxyClient;
use convertor::core::profile::Profile;
use convertor::core::profile::clash_profile::ClashProfile;
use convertor::core::profile::extract_policies_for_rule_provider;
use convertor::core::profile::policy::Policy;
use convertor::core::profile::surge_profile::SurgeProfile;
use convertor::provider_api::ProviderApi;
use convertor::url::convertor_url::{ConvertorUrl, ConvertorUrlType};
use convertor::url::query::ConvertorQuery;
use convertor::url::url_builder::{HostPort, UrlBuilder};
use convertor::url::url_error::UrlBuilderError;
use headers::UserAgent;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fmt;
use std::fmt::Display;
use url::Url;

#[derive(Default, Debug, Clone, Hash, Args)]
pub struct ProviderCmd {
    /// 构造适用于不同客户端的订阅地址
    #[arg(value_enum)]
    pub client: ProxyClient,

    /// 订阅提供商
    #[arg(value_enum, default_value_t = Provider::BosLife)]
    pub provider: Provider,

    /// 订阅链接, 可以是 转换器链接(profile_url) 或 原始订阅链接(raw_url)
    /// 将会自动判断链接类型, 仅根据 host 部分与 server 进行匹配
    /// 可选的参数, 如果不指定, 则通过 provider 来获取 raw_url
    #[arg()]
    pub url: Option<Url>,

    /// convertor 所在服务器的地址
    /// 格式为 `http://host:port`
    /// 未指定时，使用配置文件中的默认值
    #[arg(short, long)]
    pub server: Option<Url>,

    /// 订阅更新的间隔时间，单位为秒
    /// 未指定时，使用配置文件中的默认值
    #[arg(short, long)]
    pub interval: Option<u64>,

    /// 是否严格模式
    /// 如果开启，订阅转换器将严格按照配置进行转换
    #[arg(short = 'S', long)]
    pub strict: Option<bool>,

    /// 是否重置订阅链接
    #[arg(short, long, default_value_t = false)]
    pub reset: bool,

    #[cfg(feature = "update")]
    /// 是否更新本地订阅文件
    #[arg(short, long, default_value_t = false)]
    pub update: bool,
}

pub struct ProviderCli {
    pub config: ConvertorConfig,
    pub api_map: HashMap<Provider, ProviderApi>,
}

#[allow(clippy::large_enum_variant)]
enum ClientProfile {
    Surge,
    #[cfg(feature = "update")]
    Clash(ClashProfile),
    #[cfg(not(feature = "update"))]
    Clash,
}

impl ProviderCli {
    pub fn new(config: ConvertorConfig, api_map: HashMap<Provider, ProviderApi>) -> Self {
        Self { config, api_map }
    }

    pub async fn execute(&mut self, cmd: ProviderCmd) -> Result<(UrlBuilder, ProviderCliResult)> {
        let client = cmd.client;
        let provider = cmd.provider;
        let url_builder = self.create_url_builder(&cmd).await?;
        let api = self
            .api_map
            .get_mut(&provider)
            .ok_or(eyre!("无法取得订阅供应商的 api 实现: {}", &provider))?;
        api.set_sub_url(url_builder.sub_url.clone());
        let raw_profile_content = api
            .get_raw_profile(client, UserAgent::from_static("Surge Mac/8310"))
            .await?;
        let sub_host = url_builder.sub_url.host_port()?;
        let (_client_profile, policies) = match client {
            ProxyClient::Surge => {
                let mut raw_profile = SurgeProfile::parse(raw_profile_content)?;
                raw_profile.convert(&url_builder)?;
                let mut policies: Vec<Policy> = raw_profile.policy_of_rules.keys().cloned().collect();
                policies.sort();
                (ClientProfile::Surge, policies)
            }
            ProxyClient::Clash => {
                let raw_profile = ClashProfile::parse(raw_profile_content)?;
                let policies = extract_policies_for_rule_provider(&raw_profile.rules, sub_host);
                #[cfg(feature = "update")]
                {
                    (ClientProfile::Clash(raw_profile), policies)
                }
                #[cfg(not(feature = "update"))]
                {
                    (ClientProfile::Clash, policies)
                }
            }
        };

        let raw_url = url_builder.build_raw_url();
        let profile_url = url_builder.build_profile_url()?;
        let raw_profile_url = url_builder.build_raw_profile_url()?;
        let sub_logs_url = url_builder.build_sub_logs_url()?;
        let rule_provider_urls = policies
            .iter()
            .map(|policy| url_builder.build_rule_provider_url(policy))
            .collect::<Result<Vec<_>, UrlBuilderError>>()?;
        let result = ProviderCliResult {
            raw_url,
            raw_profile_url,
            profile_url,
            sub_logs_url,
            rule_provider_urls,
        };

        #[cfg(feature = "update")]
        // 副作用逻辑后置，主流程只负责数据流
        if cmd.update {
            match _client_profile {
                ClientProfile::Surge => {
                    super::update::update_surge_config(&self.config, &url_builder, &policies).await?;
                }
                ClientProfile::Clash(profile) => {
                    super::update::update_clash_config(&self.config, &url_builder, profile).await?;
                }
            }
        }
        Ok((url_builder, result))
    }

    pub fn post_execute(&self, _url_builder: UrlBuilder, result: ProviderCliResult) {
        println!("{result}");
    }

    async fn create_url_builder(&self, cmd: &ProviderCmd) -> Result<UrlBuilder> {
        let ProviderCmd {
            client,
            provider,
            url,
            server,
            interval,
            strict,
            reset,
            #[cfg(feature = "update")]
                update: _,
        } = cmd;

        // 从 client_config 中取参数的 fallback
        let client_config = self
            .config
            .clients
            .get(client)
            .ok_or(eyre!("无法取得代理客户端的配置: {}", client))?;

        let server = server.clone().unwrap_or_else(|| self.config.server.clone());
        let mut enc_secret = None;
        let mut enc_sub_url = None;
        let mut interval = interval
            .as_ref()
            .map(|i| *i)
            .unwrap_or_else(|| client_config.interval());
        let mut strict = strict.as_ref().map(|s| *s).unwrap_or_else(|| client_config.strict());

        let url_type = self.detect_url(cmd);
        let sub_url = match (url_type, reset) {
            // Get sub_url
            (None, false) => {
                self.api_map
                    .get(provider)
                    .ok_or(eyre!("无法取得订阅供应商的 api 实现: {}", &provider))?
                    .get_sub_url()
                    .await?
            }
            // Reset sub_url
            (None, true) => {
                self.api_map
                    .get(provider)
                    .ok_or(eyre!("无法取得订阅供应商的 api 实现: {}", &provider))?
                    .reset_sub_url()
                    .await?
            }
            // Use sub_url
            (Some(ConvertorUrlType::Raw), _) => url.clone().unwrap(),
            // Decode profile_url
            (Some(ConvertorUrlType::Profile), _) => {
                let profile_query = url.as_ref().and_then(Url::query).ok_or(eyre!("订阅链接缺少查询参数"))?;
                let query = ConvertorQuery::parse_from_query_string(
                    profile_query,
                    &self.config.secret,
                    server.clone(),
                    *client,
                    *provider,
                )?;
                enc_secret = query.enc_secret.clone();
                enc_sub_url = Some(query.enc_sub_url.clone());
                interval = query.interval;
                strict = query.strict.unwrap_or(strict);
                query.sub_url.clone()
            }
            _ => unreachable!("不支持的订阅链接类型"),
        };

        let url_builder = UrlBuilder::new(
            self.config.secret.clone(),
            enc_secret,
            *client,
            *provider,
            server,
            sub_url,
            enc_sub_url,
            interval,
            strict,
        )?;

        Ok(url_builder)
    }

    fn detect_url(&self, cmd: &ProviderCmd) -> Option<ConvertorUrlType> {
        let ProviderCmd { url, server, .. } = cmd;
        let server = server
            .as_ref()
            .map(|s| s.to_string())
            .unwrap_or_else(|| self.config.server.to_string());
        url.as_ref().map(|url| {
            if url.as_str().starts_with(&server) {
                ConvertorUrlType::Profile
            } else {
                ConvertorUrlType::Raw
            }
        })
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProviderCliResult {
    pub raw_url: ConvertorUrl,
    pub raw_profile_url: ConvertorUrl,
    pub profile_url: ConvertorUrl,
    pub sub_logs_url: ConvertorUrl,
    pub rule_provider_urls: Vec<ConvertorUrl>,
}

impl Display for ProviderCliResult {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "{}", self.raw_url.r#type.blue())?;
        writeln!(f, "{}", self.raw_url)?;
        writeln!(f, "{}", self.profile_url.r#type.blue())?;
        writeln!(f, "{}", self.profile_url)?;
        writeln!(f, "{}", self.raw_profile_url.r#type.blue())?;
        writeln!(f, "{}", self.raw_profile_url)?;
        writeln!(f, "{}", self.sub_logs_url.r#type.blue())?;
        writeln!(f, "{}", self.sub_logs_url)?;
        writeln!(f, "{}", ConvertorUrlType::RuleProvider.to_string().blue())?;
        for link in &self.rule_provider_urls {
            writeln!(f, "{link}")?;
        }
        Ok(())
    }
}

impl ProviderCmd {
    pub fn snapshot_name(&self) -> String {
        let client = self.client.to_string();
        let provider = self.provider.to_string();
        let url = self
            .url
            .as_ref()
            .map_or("no_url".to_string(), |url| url.host_port().unwrap());
        let server = self
            .server
            .as_ref()
            .map_or("no_server".to_string(), |server| server.to_string());
        let interval = self
            .interval
            .map_or("no_interval".to_string(), |interval| interval.to_string());
        let strict = self.strict.map_or("no_strict".to_string(), |_| "strict".to_string());
        let reset = if self.reset { "reset" } else { "no_reset" };

        #[cfg(feature = "update")]
        let update = if self.update { "update" } else { "no_update" };

        #[cfg(not(feature = "update"))]
        let update = "no_update";

        format!("{client}-{provider}-{url}-{server}-{interval}-{strict}-{reset}-{update}")
    }
}
