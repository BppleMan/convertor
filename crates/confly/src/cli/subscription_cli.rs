use crate::config::ConflyConfig;
use clap::Args;
use color_eyre::Result;
use color_eyre::eyre::OptionExt;
use convertor::common::encrypt::encrypt;
use convertor::config::proxy_client::ProxyClient;
use convertor::core::profile::Profile;
use convertor::core::profile::clash_profile::ClashProfile;
use convertor::core::profile::extract_policies_for_rule_provider;
use convertor::core::profile::policy::Policy;
use convertor::core::profile::surge_profile::SurgeProfile;
use convertor::error::UrlBuilderError;
use convertor::provider::SubscriptionProvider;
use convertor::url::url_builder::{HostPort, UrlBuilder};
use convertor::url::url_result::UrlResult;
use url::Url;

#[derive(Default, Debug, Clone, Hash, Args)]
pub struct SubscriptionCmd {
    /// 构造适用于不同客户端的订阅地址
    #[arg(value_enum)]
    pub client: ProxyClient,

    /// 原始订阅链接(raw_url)
    #[arg()]
    pub url: Option<Url>,

    /// 是否更新本地订阅文件
    #[arg(short, long, default_value_t = false)]
    pub update: bool,
}

pub struct ProviderCli {
    pub config: ConflyConfig,
    pub provider: SubscriptionProvider,
}

#[allow(clippy::large_enum_variant)]
enum ClientProfile {
    Surge,
    Clash(ClashProfile),
}

impl ProviderCli {
    pub fn new(config: ConflyConfig) -> Self {
        let provider = SubscriptionProvider::new(None);
        Self { config, provider }
    }

    pub async fn execute(&mut self, cmd: SubscriptionCmd) -> Result<(UrlBuilder, UrlResult)> {
        let client = cmd.client;
        let url_builder = self.create_url_builder(&cmd).await?;
        let raw_url = url_builder.build_raw_url();
        let raw_profile_content = self
            .provider
            .get_raw_profile(raw_url.into(), ([("User-Agent", "Surge Mac/8310")].into()))
            .await?;
        let sub_host = url_builder
            .sub_url
            .host_port()
            .ok_or_eyre("无法从 sub_url 中提取 host port")?;
        let (client_profile, policies) = match client {
            ProxyClient::Surge => {
                let mut raw_profile = SurgeProfile::parse(raw_profile_content)?;
                raw_profile.convert(&url_builder)?;
                let mut policies: Vec<Policy> = raw_profile.policy_of_rules.keys().cloned().collect();
                policies.sort();
                (ClientProfile::Surge, policies)
            }
            ProxyClient::Clash => {
                let raw_profile = ClashProfile::parse(raw_profile_content)?;
                let policies = extract_policies_for_rule_provider(&raw_profile.rules, sub_host);
                (ClientProfile::Clash(raw_profile), policies)
            }
        };

        let raw_url = url_builder.build_raw_url();
        let profile_url = url_builder.build_profile_url()?;
        let raw_profile_url = url_builder.build_raw_profile_url()?;
        let rule_provider_urls = policies
            .iter()
            .map(|policy| url_builder.build_rule_provider_url(policy))
            .collect::<Result<Vec<_>, UrlBuilderError>>()?;
        let result = UrlResult {
            raw_url,
            raw_profile_url,
            profile_url,
            rule_providers_url: rule_provider_urls,
        };

        // 副作用逻辑后置，主流程只负责数据流
        if cmd.update {
            match (client_profile, self.config.clients.get(&client)) {
                (ClientProfile::Surge, Some(client_config)) => {
                    client_config.update_surge_config(&url_builder, &policies).await?;
                }
                (ClientProfile::Clash(profile), Some(client_config)) => {
                    client_config
                        .update_clash_config(&url_builder, profile, &self.config.common.secret)
                        .await?;
                }
                _ => eprintln!("未找到对应的客户端配置，跳过更新本地订阅文件"),
            }
        }
        Ok((url_builder, result))
    }

    pub fn post_execute(&self, _url_builder: UrlBuilder, result: UrlResult) {
        println!("{result}");
    }

    async fn create_url_builder(&self, cmd: &SubscriptionCmd) -> Result<UrlBuilder> {
        let SubscriptionCmd { client, url, update: _ } = cmd;
        let subscription_config = &self.config.common.subscription;

        let sub_url = url.clone().unwrap_or(subscription_config.sub_url.clone());
        let server = self.config.common.server.clone();
        let secret = self.config.common.secret.clone();
        let enc_secret = encrypt(secret.as_bytes(), &secret)?;
        let enc_sub_url = encrypt(secret.as_bytes(), sub_url.as_str())?;
        let interval = subscription_config.interval;
        let strict = subscription_config.strict;

        let url_builder = UrlBuilder::new(
            secret,
            Some(enc_secret),
            *client,
            server,
            sub_url,
            Some(enc_sub_url),
            interval,
            strict,
        )?;

        Ok(url_builder)
    }
}

// impl SubscriptionCmd {
//     pub fn snapshot_name(&self) -> String {
//         let client = self.client.to_string();
//         let url = self
//             .url
//             .as_ref()
//             .map_or("no_url".to_string(), |url| url.host_port().unwrap());
//         let server = self
//             .c
//             .as_ref()
//             .map_or("no_server".to_string(), |server| server.to_string());
//         let interval = self
//             .interval
//             .map_or("no_interval".to_string(), |interval| interval.to_string());
//         let strict = self.strict.map_or("no_strict".to_string(), |_| "strict".to_string());
//         let reset = if self.reset { "reset" } else { "no_reset" };
//
//         let update = if self.update { "update" } else { "no_update" };
//
//         format!("{client}-{provider}-{url}-{server}-{interval}-{strict}-{reset}-{update}")
//     }
// }
