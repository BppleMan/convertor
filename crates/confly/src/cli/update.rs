use crate::config::{ClientConfig, ConflyConfig};
use color_eyre::owo_colors::OwoColorize;
use convertor::config::proxy_client::ProxyClient;
use convertor::core::profile::Profile;
use convertor::core::profile::clash_profile::ClashProfile;
use convertor::core::profile::policy::Policy;
use convertor::core::profile::rule::Rule;
use convertor::core::profile::surge_header::SurgeHeader;
use convertor::core::renderer::Renderer;
use convertor::core::renderer::clash_renderer::ClashRenderer;
use convertor::core::renderer::surge_renderer::{
    SURGE_RULE_PROVIDER_COMMENT_END, SURGE_RULE_PROVIDER_COMMENT_START, SurgeRenderer,
};
use convertor::url::convertor_url::{ConvertorUrl, ConvertorUrlType};
use convertor::url::url_builder::UrlBuilder;
use std::borrow::Cow;
use std::path::Path;
// pub async fn update_surge_config(
//     config: &ConflyConfig,
//     url_builder: &UrlBuilder,
//     policies: impl IntoIterator<Item = &Policy>,
// ) -> color_eyre::Result<()> {
//     if let Some(client_config) = config.clients.get(&ProxyClient::Surge) {
//         surge_config.update_surge_config(url_builder, policies).await?;
//     } else {
//         eprintln!("{}", "Surge 配置未找到，请检查配置文件是否正确设置".red().bold());
//     }
//     Ok(())
// }
//
// pub async fn update_clash_config(
//     config: &ConvertorConfig,
//     url_builder: &UrlBuilder,
//     raw_profile: ClashProfile,
// ) -> color_eyre::Result<()> {
//     if let Some(ProxyClientConfig::Clash(clash_config)) = config.clients.get(&ProxyClient::Clash) {
//         clash_config
//             .update_clash_config(url_builder, raw_profile, &config.secret)
//             .await?;
//     } else {
//         eprintln!("{}", "Clash 配置未找到，请检查配置文件是否正确设置".red().bold());
//     }
//     Ok(())
// }

impl ClientConfig {
    pub async fn update_surge_config(
        &self,
        url_builder: &UrlBuilder,
        policies: impl IntoIterator<Item = &Policy>,
    ) -> color_eyre::Result<()> {
        // 更新主订阅配置，即由 convertor 生成的订阅配置
        Self::update_conf(
            &self.main_profile_path(),
            url_builder.build_surge_header(ConvertorUrlType::Profile)?,
        )
        .await?;

        // 更新转发原始订阅配置，即由 convertor 生成的原始订阅配置
        if let Some(raw_profile_path) = self.raw_profile_path() {
            Self::update_conf(
                raw_profile_path,
                url_builder.build_surge_header(ConvertorUrlType::RawProfile)?,
            )
            .await?;
        }

        // 更新原始订阅配置，即由订阅提供商生成的订阅配置，如果存在的话
        if let Some(raw_sub_path) = self.raw_sub_path() {
            Self::update_conf(raw_sub_path, url_builder.build_surge_header(ConvertorUrlType::Raw)?).await?;
        }

        // 更新 rules.dconf 中的 RULE-SET 规则，规则提供者将从 policies 中生成 URL
        if let Some(rules_path) = self.rules_path() {
            self.update_surge_rule_providers(rules_path, url_builder, policies)
                .await?;
        }

        Ok(())
    }

    async fn update_surge_rule_providers(
        &self,
        rules_path: impl AsRef<Path>,
        url_builder: &UrlBuilder,
        policies: impl IntoIterator<Item = &Policy>,
    ) -> color_eyre::Result<()> {
        let content = tokio::fs::read_to_string(&rules_path).await?;
        let mut lines = content.lines().map(Cow::Borrowed).collect::<Vec<_>>();

        let range_of_rule_providers = lines.iter().enumerate().fold(0..=0, |acc, (no, line)| {
            let mut start = *acc.start();
            let mut end = *acc.end();
            if line == SURGE_RULE_PROVIDER_COMMENT_START {
                start = no;
            } else if line == SURGE_RULE_PROVIDER_COMMENT_END {
                end = no;
            }
            start..=end
        });

        let provider_rules = policies
            .into_iter()
            .map(|policy| {
                let name = SurgeRenderer::render_provider_name_for_policy(policy);
                let url = url_builder.build_rule_provider_url(policy)?;
                Ok(Rule::surge_rule_provider(policy, name, url))
            })
            .collect::<color_eyre::Result<Vec<_>>>()?;
        let mut output = provider_rules
            .iter()
            .map(SurgeRenderer::render_rule)
            .map(|l| Ok(l.map(Cow::Owned)?))
            .collect::<color_eyre::Result<Vec<_>>>()?;
        output.insert(0, Cow::Borrowed(SURGE_RULE_PROVIDER_COMMENT_START));
        output.push(Cow::Borrowed(SURGE_RULE_PROVIDER_COMMENT_END));
        lines.splice(range_of_rule_providers, output);
        let content = lines.join("\n");
        tokio::fs::write(rules_path, &content).await?;
        Ok(())
    }

    async fn update_conf(config_path: impl AsRef<Path>, header: SurgeHeader) -> color_eyre::Result<()> {
        let mut content = tokio::fs::read_to_string(&config_path).await?;
        let mut lines = content.lines().map(Cow::Borrowed).collect::<Vec<_>>();
        lines[0] = Cow::Owned(header.to_string());
        content = lines.join("\n");
        tokio::fs::write(&config_path, &content).await?;
        Ok(())
    }

    pub async fn update_clash_config(
        &self,
        url_builder: &UrlBuilder,
        raw_profile: ClashProfile,
        secret: impl AsRef<str>,
    ) -> color_eyre::Result<()> {
        let mut template = ClashProfile::template()?;
        template.patch(raw_profile)?;
        template.convert(url_builder)?;
        template.secret = Some(secret.as_ref().to_string());
        let clash_config = ClashRenderer::render_profile(&template)?;
        let main_sub_path = self.main_profile_path();
        if !main_sub_path.is_file() {
            if let Some(parent) = main_sub_path.parent() {
                tokio::fs::create_dir_all(parent).await?;
            }
        }
        tokio::fs::write(main_sub_path, clash_config).await?;
        Ok(())
    }
}
