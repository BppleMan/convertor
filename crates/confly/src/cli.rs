use crate::cli::config_cli::ConfigCmd;
use crate::cli::subscription_cli::SubscriptionCmd;
use clap::Subcommand;

pub mod config_cli;
pub mod subscription_cli;
pub mod update;

#[derive(Debug, Clone, Subcommand)]
#[allow(clippy::large_enum_variant)]
pub enum ConvertorCommand {
    /// 配置相关的子命令
    /// 获取配置模板, 生成配置文件等
    #[command(name = "config")]
    Config(ConfigCmd),

    /// 获取订阅提供商的订阅链接
    #[command(name = "sub")]
    Subscription(SubscriptionCmd),
}
