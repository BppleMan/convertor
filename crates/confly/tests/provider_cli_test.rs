use color_eyre::Report;
use color_eyre::eyre::eyre;
use convertor::common::config::ConvertorConfig;
use convertor::common::config::proxy_client_config::ProxyClient;
use convertor::provider_api::ProviderApi;

use confly::cli::provider_cli::{ProviderCli, ProviderCmd};
use convertor::common::config::provider_config::Provider;
use convertor::testkit::{init_test, start_mock_provider_server};
use convertor::url::url_builder::HostPort;
use pretty_assertions::assert_str_eq;
use regex::Regex;
use url::Url;

pub fn cmds(client: ProxyClient, provider: Provider) -> [ProviderCmd; 3] {
    [
        ProviderCmd {
            client,
            provider,
            ..Default::default()
        },
        ProviderCmd {
            client,
            provider,
            server: Some(Url::parse("http://localhost:8080").expect("不合法的服务器地址")),
            ..Default::default()
        },
        ProviderCmd {
            client,
            provider,
            interval: Some(43200),
            strict: Some(false),
            ..Default::default()
        },
    ]
}

async fn test_subscription(
    convertor_config: &ConvertorConfig,
    client: ProxyClient,
    provider: Provider,
    cmd: ProviderCmd,
) -> Result<(), Report> {
    let convertor_config = convertor_config.clone();
    let server = match &cmd.server {
        Some(server) => server.clone(),
        None => convertor_config.server.clone(),
    };
    let client_config = convertor_config
        .clients
        .get(&client)
        .ok_or_else(|| eyre!("未找到客户端配置: {}", client))?;
    let interval = cmd.interval.unwrap_or(client_config.interval());
    let strict = cmd.strict.unwrap_or(client_config.strict());

    let api_map = ProviderApi::create_api_no_redis(convertor_config.providers.clone());
    let mut executor = ProviderCli::new(convertor_config, api_map);
    let (url_builder, result) = executor.execute(cmd).await?;

    // 构造期望值
    let expect_raw_url = format!(
        "{}://{}/subscription?token=bppleman&flag={}",
        url_builder.sub_url.scheme(),
        url_builder.sub_url.host_port()?,
        client
    );
    assert_str_eq!(expect_raw_url, result.raw_url.to_string());

    let expect_profile_url = format!(
        "{server}profile?client={client}&provider={provider}&server={server}&interval={interval}&strict={strict}&sub_url={}",
        url_builder.enc_sub_url
    );
    assert_str_eq!(expect_profile_url, result.profile_url.to_string());

    let regex_str = format!(
        r#"{}sub-logs\?provider={provider}&secret=(?P<enc_secret>.+)&page=1&page_size=20"#,
        server.as_str().replace(".", "\\."),
    );
    let regex = Regex::new(&regex_str)?;
    let sub_logs_url_string = result.sub_logs_url.to_string();
    let Some(captures) = regex.captures(&sub_logs_url_string) else {
        panic!("未正确捕获订阅日志链接");
    };
    let Some(enc_secret) = captures.name("enc_secret") else {
        panic!("未正确捕获 enc_secret");
    };
    executor.config.validate_enc_secret(enc_secret.as_str())?;

    let expect_policy_urls = [
        format!(
            "{server}rule-provider?client={client}&provider={provider}&server={server}&interval={interval}&policy.name=DIRECT&policy.is_subscription=true&sub_url={}",
            url_builder.enc_sub_url
        ),
        format!(
            "{server}rule-provider?client={client}&provider={provider}&server={server}&interval={interval}&policy.name=BosLife&policy.is_subscription=false&sub_url={}",
            url_builder.enc_sub_url
        ),
        format!(
            "{server}rule-provider?client={client}&provider={provider}&server={server}&interval={interval}&policy.name=BosLife&policy.option=no-resolve&policy.is_subscription=false&sub_url={}",
            url_builder.enc_sub_url
        ),
        format!(
            "{server}rule-provider?client={client}&provider={provider}&server={server}&interval={interval}&policy.name=BosLife&policy.option=force-remote-dns&policy.is_subscription=false&sub_url={}",
            url_builder.enc_sub_url
        ),
        format!(
            "{server}rule-provider?client={client}&provider={provider}&server={server}&interval={interval}&policy.name=DIRECT&policy.is_subscription=false&sub_url={}",
            url_builder.enc_sub_url
        ),
        format!(
            "{server}rule-provider?client={client}&provider={provider}&server={server}&interval={interval}&policy.name=DIRECT&policy.option=no-resolve&policy.is_subscription=false&sub_url={}",
            url_builder.enc_sub_url
        ),
        format!(
            "{server}rule-provider?client={client}&provider={provider}&server={server}&interval={interval}&policy.name=DIRECT&policy.option=force-remote-dns&policy.is_subscription=false&sub_url={}",
            url_builder.enc_sub_url
        ),
    ];

    for (i, url) in result.rule_provider_urls.iter().enumerate() {
        assert_str_eq!(expect_policy_urls[i], url.to_string());
    }

    Ok(())
}

#[tokio::test]
async fn test_subscription_surge_boslife() -> color_eyre::Result<()> {
    init_test();
    let mut config = ConvertorConfig::template();
    config.providers.values_mut().for_each(|provider| {
        provider.api_config.headers.remove("Authorization");
    });
    // {
    //     调试
    //     let api_map = ProviderApi::create_api_no_redis(config.providers.clone());
    //     let provider_api = api_map
    //         .get(&Provider::BosLife)
    //         .ok_or_else(|| eyre!("未找到 BosLife 提供者"))?;
    //     provider_api.login().await?;
    // }
    start_mock_provider_server(&mut config).await?;
    let client = ProxyClient::Surge;
    let provider = Provider::BosLife;
    let cmds = cmds(client, provider);
    for cmd in cmds {
        test_subscription(&config, client, provider, cmd).await?;
    }
    Ok(())
}

// #[tokio::test]
// async fn test_subscription_clash_boslife() -> color_eyre::Result<()> {
//     let convertor_config = convertor_config().await;
//     let client = ProxyClient::Clash;
//     let provider = Provider::BosLife;
//     let cmds = cmds(client, provider);
//     for cmd in cmds {
//         test_subscription(&convertor_config, client, provider, cmd).await?;
//     }
//     Ok(())
// }
