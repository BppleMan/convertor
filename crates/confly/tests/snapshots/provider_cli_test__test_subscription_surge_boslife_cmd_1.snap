---
source: crates/confly/tests/provider_cli_test.rs
expression: result
---
[34m提供商原始订阅[39m
http://127.0.0.1:<PORT>/subscription?token=bppleman&flag=Surge
[34m转换器订阅[39m
<SERVER>profile/surge/boslife?interval=43200&strict=true&sub_url=<ENC_SUB_URL>
[34m转发原始订阅[39m
<SERVER>raw-profile/surge/boslife?interval=43200&strict=true&sub_url=<ENC_SUB_URL>
[34m订阅日志记录[39m
<SERVER>api/sub-logs?secret=akO49BUYoRzDh7Zpsu5lhp8H575VUTh6Mh74Xh0JSf5jblvyCnUxdxwp2ClcBhAf
[34m规则集: [Subscription][39m
<SERVER>rule-provider/surge/boslife?interval=43200&policy[name]=DIRECT&policy[is_subscription]=true&sub_url=<ENC_SUB_URL>
[34m规则集: [BosLife][39m
<SERVER>rule-provider/surge/boslife?interval=43200&policy[name]=BosLife&policy[is_subscription]=false&sub_url=<ENC_SUB_URL>
[34m规则集: [BosLife: no-resolve][39m
<SERVER>rule-provider/surge/boslife?interval=43200&policy[name]=BosLife&policy[option]=no-resolve&policy[is_subscription]=false&sub_url=<ENC_SUB_URL>
[34m规则集: [BosLife: force-remote-dns][39m
<SERVER>rule-provider/surge/boslife?interval=43200&policy[name]=BosLife&policy[option]=force-remote-dns&policy[is_subscription]=false&sub_url=<ENC_SUB_URL>
[34m规则集: [DIRECT][39m
<SERVER>rule-provider/surge/boslife?interval=43200&policy[name]=DIRECT&policy[is_subscription]=false&sub_url=<ENC_SUB_URL>
[34m规则集: [DIRECT: no-resolve][39m
<SERVER>rule-provider/surge/boslife?interval=43200&policy[name]=DIRECT&policy[option]=no-resolve&policy[is_subscription]=false&sub_url=<ENC_SUB_URL>
[34m规则集: [DIRECT: force-remote-dns][39m
<SERVER>rule-provider/surge/boslife?interval=43200&policy[name]=DIRECT&policy[option]=force-remote-dns&policy[is_subscription]=false&sub_url=<ENC_SUB_URL>
